rustls/src/server/mod.rs

```diff
@@ -146,6 +146,311 @@ impl<'a> ClientHello<'a> {
+/// 2. You now have a ServerConfigBuilderWithSuites.  You must make a decision
+///    on key exchange groups: typically by calling `with_default_kx_groups()`.
+/// 2. You now have a ServerConfigBuilderWithSuitesAndKx.  You must make
+///    a decision on how and whether to use client authentication.  Perhaps
+///    you call `without_client_auth()`.
+/// 3. You now have a ServerConfigBuilderComplete.  This object has a number
+///    of defaults you can change, or you can pass it to `ServerConfig::new`.
+pub struct ServerConfigBuilder {}
+
+impl ServerConfigBuilder {
+    /// Start building a ServerConfig.
+    pub fn new() -> ServerConfigBuilder {
+        ServerConfigBuilder {}
+    }
+
+    /// Start building a ServerConfig, and accept defaults for underlying
+    /// cryptography.
+    ///
+    /// These are safe defaults, useful for 99% of applications.
+    ///
+    /// With the returned object, you must still make decisions on
+    /// client authentication and provide server credentials -- rustls
+    /// can't provide useful defaults for these.
+    pub fn defaults() -> ServerConfigBuilderWithSuitesAndKxGroups {
+        ServerConfigBuilder::new()
+            .with_default_ciphersuites()
+            .with_default_kx_groups()
+    }
+
+    /// Choose a specific set of ciphersuites.
+    pub fn with_ciphersuites(
+        self,
+        ciphersuites: &[&'static SupportedCipherSuite],
+    ) -> ServerConfigBuilderWithSuites {
+        ServerConfigBuilderWithSuites {
+            ciphersuites: ciphersuites.to_vec(),
+        }
+    }
+
+    /// Choose the default set of ciphersuites.
+    ///
+    /// Note that this default provides only high-quality suites: there is no need
+    /// to filter out low-, export- or NULL-strength ciphersuites: rustls does not
+    /// implement these.
+    pub fn with_default_ciphersuites(self) -> ServerConfigBuilderWithSuites {
+        self.with_ciphersuites(DEFAULT_CIPHERSUITES)
+    }
+}
+
+/// A ServerConfigBuilder where we know the ciphersuites.
+pub struct ServerConfigBuilderWithSuites {
+    ciphersuites: Vec<&'static SupportedCipherSuite>,
+}
+
+impl ServerConfigBuilderWithSuites {
+    /// Choose a specific set of key exchange groups.
+    pub fn with_kx_groups(
+        self,
+        kx_groups: &[&'static SupportedKxGroup],
+    ) -> ServerConfigBuilderWithSuitesAndKxGroups {
+        ServerConfigBuilderWithSuitesAndKxGroups {
+            ciphersuites: self.ciphersuites,
+            kx_groups: kx_groups.to_vec(),
+        }
+    }
+
+    /// Choose the default set of key exchange groups.
+    ///
+    /// This is a safe default: rustls doesn't implement any poor-quality groups.
+    pub fn with_default_kx_groups(self) -> ServerConfigBuilderWithSuitesAndKxGroups {
+        self.with_kx_groups(&ALL_KX_GROUPS)
+    }
+}
+
+/// A ServerConfigBuilder where we know the ciphersuites and key exchange
+/// groups.
+pub struct ServerConfigBuilderWithSuitesAndKxGroups {
```

Comment:
Did you consider giving this type the `ServerConfigBuilder` name, and using some other name for the completely empty `ServerConfigBuilder`? That would somewhat alleviate my concerns about the lack of defaults. (And I think naming that type would be an interesting exercise.)

Reply:
Reflecting upon this, I think you set it up this way to prevent linking all the cipher suites even if not configured? I think we might achieve the same goal if we use something like the complete confit builder which has the cipher suites value as an Option internally which remains None until set and only gets filled in with the default (all suites) when build() is called.

Reply:
The problem with this is that build() cannot have any link-time dependency on the defaults, even a conditional one.

Reply:
Do you agree that having all cipher suites enabled should be the default and the easiest to build, and the link-time restricted smaller subset should be possible but can be slightly less ergonomic?

Maybe we have a `ServerConfigBuilder` with `new()` starts with an empty `cipher_suites: Vec<_>` and `default()` that starts with a complete `cipher_suites`, with a method `cipher_suites(&[_])` that resets the internal cipher suites? That would (a) have a convenient path towards full cipher suites with `default()`, (b) straightforward path to custom cipher suite configurations with `new().cipher_suites()`. A potential downside is that doing `default().cipher_suites()` is possible and works but is not as a efficient as it could be, but it would reduce the complexity of the three-level builder stack quite a bit, and we could warn about it in the configuration. (I think this scheme would satisfy the lack of link-time dependency goal?)

Reply:
> Do you agree that having all cipher suites enabled should be the default and the easiest to build, and the link-time restricted smaller subset should be possible but can be slightly less ergonomic?

We should distinguish between "all cipher suites" and "default cipher suites." Currently Rustls enables all suites it supports by default because it doesn't implement any sub-optimal (in our estimation) cipher suites. However, this is unlikely to be the case in the very long term.

Reply:
Yeah, I think I now better understand why this makes sense.